Step4 리뷰 요청 드립니다.

lcjltj opened the pull request:
안녕하세요!!

매번 꼼꼼한 피드백 감사드립니다.!!

최대한 실수 없이 하려고 하긴 하는데 .. 혼자 너무 진도를 못나간거같아.. 성격이 급해지네요 ㅠㅠ 

매번 감사드립니다.!! 사소한 것이라도 피드백 부탁드립니다.

리팩토링을하면서 TDD로 개발 하면 수정하기 편하다는걸 좀 느낀거같아요!!

lcjltj committed:
test(LottoTest): Arguments 활용한 로또 테스트 방식 변경

lcjltj committed:
docs: 보너스 당첨 번호 기능 추가

lcjltj committed:
feat(WinningNumbers): 당첨 번호 / 보너스 번호 클래스 추가

- validate() 당첨번호와 보너스 번호 중복 값 확인
- bonus() 로또 번호에 보너스 번호가 포함 되는지 확인

lcjltj committed:
feat: 보너스 당첨 추가 외 6

- Winners
    - RANK 배열 Prize.BONUS_PLACE 추가
- LottoNumber
    - ofString 팩토리 메소드 추가
- LottoExceptionType
    - DUPLICATE_BONUS_NUMBER 추가
- InputView
    - 보너스 볼 입력 메소드 추가
- ResultView
    - 보너스 볼 출력 메시지 추가
- Application, LottoController
    - exchangePrize() WinningNumbers 파라메터 변경

lcjltj committed:
refactor(Prize) : 보너스 번호 추가

- isBonus() 보너스 번호가 로또 번호에 포함 되는지 확인

lcjltj committed:
refactor: 보너스 당첨 확인 로직 이동

- Lotto
    - contains() 접근자 변경 private -> public
- Prize
    - 보너스 확인 logic 변경 Lotto 클래스 contains 사용
- WinningNumbers
    - bonus 메소드 삭제
    - BonusNumber getter 추가

parkeeseul commented on src/main/java/lotto/Application.java:
```diff
@@ -15,8 +17,9 @@ public static void main(String[] args) {
-        long totalPrize = lottoController.exchangePrize(lottoes, answer);
+        Lotto winner = new Lotto(InputView.inputAnswer());
+        LottoNumber bonusNumber = LottoNumber.ofString(InputView.inputBonusAnswer());
+        long totalPrize = lottoController.exchangePrize(lottoes, new WinningNumbers(winner, bonusNumber));
```
Controller 의 역할은 무엇인가요?
Application 클래스와는 어떠한 차이가 있나요? 😄

lcjltj replied:
Controller 는 사용자 요청에 대한 처리.

Application 은 사용자 요청하는 클래스로 생각하면서 작성 했었습니다.

하지만 application 은 하나의 프로그램을 실행시킨다는 의미로 해당 logic관련하여 controller로 이동하였습니다.!

parkeeseul replied:
Controller 는 뷰와 도메인의 연결점이 될 것 같은데요,
Application 도 비슷한 역할을 하는 것 같아서 어떠한 차이가 있는지 질문 드렸습니다 😄

parkeeseul commented on src/main/java/lotto/domain/Prize.java:
```diff
@@ -27,11 +30,13 @@ public enum Prize {
+    public static Winners matchLotto(final List<Lotto> lottoes, final WinningNumbers winningNumber) {
         return new Winners(lottoes.stream()
-                .map(v -> v.matchCount(winnigNumber))
-                .map(Prize::createPrize)
-                .collect(Collectors.toList()));
+                .map(lotto -> {
+                    long count = lotto.matchCount(winningNumber.getWinner());
+                    boolean bonus = lotto.contains(winningNumber.getBonusNumber());
+                    return createPrize(count, bonus);
+                }).collect(Collectors.toCollection(ArrayList::new)));
     }
```
해당 부분 가독성이 조금 떨어지는 것 같은데 메서드로 분리해 보면 어떨까요? 😄

lcjltj replied:
해당 메소드 분리하여 반영하였습니다.

parkeeseul commented on src/main/java/lotto/domain/Prize.java:
```diff
@@ -42,8 +47,14 @@ public long getPrize() {
     }
 
-    private static Prize createPrize(final long count) {
+    private static Prize createPrize(final long count, final boolean bonus) {
+        if (isBonus(count, bonus)) {
+            return Prize.BONUS_PLACE;
+        }
         return prizeMap.getOrDefault(Long.valueOf(count).intValue(), FAIL);
     }
```
만약 당첨 조건들이 달라지게 되면 if 문들이 추가되거나 해당 로직들이 변경될 수 있을 것 같은데요,
Prize 에 보너스볼에 대한 매칭 여부를 추가해서 당첨 조건에 대한 로직을 개선해 보는 것은 어떻게 생각하시나요? 😄

lcjltj replied:
해당 로직 values . findAny 방식으로 변경하였습니다. ㅠㅠ
기존 방식 으로 하게 되면 시간복잡도는 O(1) 이고  (배열 , 해쉬맵)
바뀐 방식은 시간 복잡도가 O(n)이 되게 되는데.. 이런 내용 관련에 대해선 어떻게 생각하시나요??

parkeeseul replied:
엄청난 성능차이가 나는 복잡한 로직이라면 고민을 해봐야 할 것 같지만
현재 미션에서는 유의미한 성능 차이가 없을 것 같아요. 😄 
이런 부분에서는 코드의 가독성과 유지보수의 이점을 가져가는 것이 더 좋다고 생각합니다 😄

parkeeseul commented on src/main/java/lotto/domain/WinningNumbers.java:
```diff
@@ -0,0 +1,30 @@
+        this.bonusNumber =  bonusNumber;
+    }
+
+    private void validate(final Lotto winner, final LottoNumber bonusNumber) {
+        if (winner.getLottoNumber().contains(bonusNumber)) {
+            throw new LottoException(LottoExceptionType.DUPLICATE_BONUS_NUMBER);
+        }
+    }
```
Lotto 의 contains 메서드를 활용해 보면 좋을 것 같아요 😄 

```suggestion
    private void validate(final Lotto winner, final LottoNumber bonusNumber) {
        if (winner.contains(bonusNumber)) {
            throw new LottoException(LottoExceptionType.DUPLICATE_BONUS_NUMBER);
        }
    }
```

lcjltj replied:
수정하였습니다.

parkeeseul commented on src/test/java/lotto/domain/WinningNumbersTest.java:
```diff
@@ -0,0 +1,20 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+class WinningNumbersTest {
```
성공하는 테스트 케이스도 작성해 보면 어떨까요? 😄

parkeeseul reviewed (changes requested):
4단계 미션 잘 진행해 주셨네요! 👍 
몇 가지 코멘트 남겨두었으니 확인 부탁드립니다 😄 
좋은 밤 되세요! 🔥

lcjltj committed:
refactor(WinningNumbers): validate contains변경 외 1

- WinningNumbersTest - 생성 테스트 작성

lcjltj committed:
refactor(Prize): bonus field 추가 외 1

- bonus ball 유무 관련 bonus 필 추가
- matchLotto
    - matchLotto() 메소드 변경 - 로또를 List로 받아 처리 하지 않고 한개의 로또만 처리하도록 변경
    - createPrize() static map을 사용 하게 되면 bonus 값을 알수 없어 values findAny 방식으로 변경

lcjltj committed:
refactor: Application 로직 controller로 이동

- 기존 Application 클래스를 사용자 관점으로 생각 하여 로직을 작성 했지만 application은 하나의 프로그램을
  실행시키는 클래스로 생각하여 로직 변경

parkeeseul reviewed (approved):
피드백 잘 반영해 주셨네요! 👍 
질문 주신 사항에 대한 저의 의견 남겨두었으니 확인 부탁드립니다 😄 
다음 단계 진행해 주세요! 화이팅 입니다 🔥
